File: android/app/src/main/java/com/dontworry/app/ui/search/SearchViewModel.kt
package com.dontworry.app.ui.search

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.dontworry.app.data.repo.ThreadRepository
import com.dontworry.app.data.repo.SynonymsRepository
import com.dontworry.app.domain.search.SearchService
import com.dontworry.app.domain.thread.ThreadIdentity
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class SearchViewModel(application: Application) : AndroidViewModel(application) {
    companion object {
        private const val MAX_RESULTS = 100
    }

    private val _uiState = MutableStateFlow(SearchUiState())
    val uiState: StateFlow<SearchUiState> = _uiState.asStateFlow()

    private var searchService: SearchService? = null

    init {
        loadData()
    }

    fun onQueryChanged(newValue: String) {
        _uiState.value = _uiState.value.copy(
            query = newValue,
            promptMessage = null,
            currentPage = 1
        )
    }

    fun onPageSelected(page: Int) {
        val state = _uiState.value
        if (state.totalPages == 0) return
        val safePage = page.coerceIn(1, state.totalPages)
        if (safePage != state.currentPage) {
            _uiState.value = state.copy(currentPage = safePage)
        }
    }

    fun submitSearch() {
        val query = _uiState.value.query.trim()
        if (query.isBlank()) {
            _uiState.value = _uiState.value.copy(promptMessage = "Please enter a keyword.")
            return
        }

        val service = searchService ?: run {
            _uiState.value = _uiState.value.copy(promptMessage = "Dataset is still loading.")
            return
        }

        _uiState.value = _uiState.value.copy(isLoading = true)

        viewModelScope.launch(Dispatchers.Default) {
            val results = service.search(query, limit = MAX_RESULTS).map {
                SearchListItem(
                    identity = ThreadIdentity.fromParts(
                        threadId = it.thread.threadId,
                        threadLink = it.thread.threadLink,
                        threadTitle = it.thread.threadTitle
                    ),
                    title = it.thread.threadTitle,
                    excerpt = it.excerpt,
                    threadLink = it.thread.threadLink,
                    threadContent = it.thread.threadContent,
                    responses = it.thread.responses
                )
            }

            _uiState.value = _uiState.value.copy(
                isLoading = false,
                promptMessage = if (results.isEmpty()) "No results found." else null,
                results = results,
                currentPage = 1
            )
        }
    }

    private fun loadData() {
        _uiState.value = _uiState.value.copy(isLoading = true)
        viewModelScope.launch(Dispatchers.Default) {
            val threads = ThreadRepository(getApplication()).loadThreads()
            val synonymsResult = SynonymsRepository(getApplication()).load()
            searchService = SearchService(threads, synonymsResult.groups)
            _uiState.value = _uiState.value.copy(
                isLoading = false,
                promptMessage = if (synonymsResult.enabled) {
                    "Type a keyword to search."
                } else {
                    "Type a keyword to search. (Synonyms disabled)"
                }
            )
        }
    }
}

class SearchViewModelFactory(
    private val application: Application
) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(SearchViewModel::class.java)) {
            @Suppress("UNCHECKED_CAST")
            return SearchViewModel(application) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class: ${modelClass.name}")
    }
}

File: ios/DontWorryiOS/Presentation/SearchViewModel.swift
import Foundation

@MainActor
final class SearchViewModel: ObservableObject {
    nonisolated private static let maxResults = 100

    @Published var query: String = ""
    @Published var isLoading = false
    @Published var isRefreshingSuggestions = false
    @Published var promptMessage: String?
    @Published var suggestedThreads: [SearchListItem] = []
    @Published var results: [SearchListItem] = []
    @Published var currentPage = 1

    let pageSize = 10

    private var allThreads: [Thread] = []
    private var searchService: SearchService?
    private var hasLoadedInitialData = false

    var totalPages: Int {
        if results.isEmpty { return 0 }
        return ((results.count - 1) / pageSize) + 1
    }

    var currentPageResults: [SearchListItem] {
        guard !results.isEmpty else { return [] }
        let safePage = min(max(currentPage, 1), totalPages)
        let from = (safePage - 1) * pageSize
        let to = min(from + pageSize, results.count)
        return Array(results[from..<to])
    }

    func loadData() {
        if isLoading || hasLoadedInitialData { return }
        hasLoadedInitialData = true
        isLoading = true

        Task.detached(priority: .userInitiated) {
            let threads = ThreadRepository().loadThreads()
            let synonyms = SynonymsRepository().load()
            let service = SearchService(threads: threads, synonymGroups: synonyms.groups)
            let suggested = threads.shuffled().prefix(10).map { Self.toSearchListItem($0) }

            await MainActor.run {
                self.allThreads = threads
                self.searchService = service
                self.suggestedThreads = Array(suggested)
                self.isLoading = false
                if threads.isEmpty {
                    self.hasLoadedInitialData = false
                    self.promptMessage = "Khong tai duoc du lieu tim kiem. Vui long thu dong app va mo lai."
                } else {
                    self.promptMessage = synonyms.enabled
                        ? "Tra cứu các kiến thức thực tế về các bệnh STDs.\nDữ liệu được cập nhật liên tục từ: diendanhiv.vn\nAnh Tuanmecsedec là admin của diễn đàn với hơn 18 năm kinh nghiệm tư vấn có chứng chỉ chuyên môn."
                        : "Tra cứu các kiến thức thực tế về các bệnh STDs.\nDữ liệu được cập nhật liên tục từ: diendanhiv.vn"
                }
            }
        }
    }

    func onQueryChanged(_ newValue: String) {
        guard newValue != query else { return }
        query = newValue
        promptMessage = nil
        currentPage = 1
    }

    func submitSearch() {
        let normalized = query.trimmingCharacters(in: .whitespacesAndNewlines)
        if normalized.isEmpty {
            promptMessage = "Hãy nhập từ khóa để tìm kiếm."
            return
        }

        guard let service = searchService else {
            promptMessage = "Dataset is still loading."
            return
        }

        isLoading = true
        let maxResults = Self.maxResults

        Task.detached(priority: .userInitiated) {
            let mapped = service.search(query: normalized, limit: maxResults).map {
                Self.toSearchListItem($0.thread, excerpt: $0.excerpt)
            }

            await MainActor.run {
                self.results = mapped
                self.currentPage = 1
                self.promptMessage = mapped.isEmpty ? "No results found." : nil
                self.isLoading = false
            }
        }
    }

    func onPageSelected(_ page: Int) {
        guard totalPages > 0 else { return }
        let safe = min(max(page, 1), totalPages)
        if safe != currentPage {
            currentPage = safe
        }
    }

    func refreshSuggestedThreads() {
        guard !allThreads.isEmpty else { return }
        isRefreshingSuggestions = true
        let threads = allThreads

        Task.detached(priority: .utility) {
            let newSuggested = threads.shuffled().prefix(10).map { Self.toSearchListItem($0) }
            await MainActor.run {
                self.suggestedThreads = Array(newSuggested)
                self.isRefreshingSuggestions = false
            }
        }
    }

    nonisolated static func toSearchListItem(_ thread: Thread, excerpt: String? = nil) -> SearchListItem {
        SearchListItem(
            id: ThreadIdentity.fromParts(
                threadId: thread.threadId,
                threadLink: thread.threadLink,
                threadTitle: thread.threadTitle
            ),
            title: thread.threadTitle,
            excerpt: excerpt ?? ExcerptBuilder.fromText(thread.threadContent),
            threadLink: thread.threadLink,
            threadContent: thread.threadContent,
            responses: thread.responses
        )
    }
}
